File: Problems/DailyTemperatures.java
/**
 * Given an array of integers temperatures represents the daily temperatures, 
 * return an array answer such that answer[i] is the number of days you have to wait 
 * after the ith day to get a warmer temperature. 
 * If there is no future day for which this is possible, keep answer[i] == 0 instead.
 */
package Problems;

import java.util.*;

public class DailyTemperatures {
    // a solution using a stack
    // time complexity O(n)
    // space complexity O(n)
    public static int[] dailyTemperatures(int[] temperatures) {
        if (temperatures.length == 1)
            return new int[] {0};
        
        Stack<Integer> stack = new Stack<>();
        int[] res = new int[temperatures.length];
        for (int i = 1; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int index = stack.pop();
                res[index] = i - index;
            }
            if (temperatures[i] > temperatures[i - 1]) {
                res[i - 1] = 1;
            } else {
                stack.add(i - 1);
            }
        }
        return res;
    }

    // another solution without using a stack
    // time complexity O(n)
    // space complexity O(2n)
    public static int[] anotherDailyTemperatures(int[] temperatures) {
        if (temperatures.length == 1)
            return new int[] {0};
        
        int[] storage = new int[temperatures.length];
        int[] res = new int[temperatures.length];
        int endOfStorage = -1;
        for (int i = 1; i < temperatures.length; i++) {
            while (endOfStorage >= 0 && temperatures[storage[endOfStorage]] < temperatures[i]) {
                int index = storage[endOfStorage--];
                res[index] = i - index;
            }
            if (temperatures[i] > temperatures[i - 1]) {
                res[i - 1] = 1;
            } else {
                storage[++endOfStorage] = i - 1;
            }
        }
        return res;
    }
    
    public static void main(String[] args) {
        int[] temperatures = { 89, 62, 70, 58, 47, 47, 46, 76, 100, 70 };
        int[] res = anotherDailyTemperatures(temperatures);
        for (int i : res) {
            System.out.print(i + " ");
        }
    }
}

File: Problems/DailyTemperatures.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Solution {
   public:  // solution using priority queue (min heap); time complexity: O(nlogn)
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        if (temperatures.size() == 1)
            return {0};
        vector<int> res(temperatures.size(), 0);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        for (int i = 1; i < temperatures.size(); i++) {
            if (temperatures[i - 1] < temperatures[i])
                res[i - 1] = 1;
            else
                minHeap.push(pair<int, int>(temperatures[i - 1], i - 1));
            while (!minHeap.empty() && minHeap.top().first < temperatures[i]) {
                auto p = minHeap.top();
                minHeap.pop();
                res[p.second] = i - p.second;
            }
        }
        return res;
    }

    // O(n) time complexity and O(1) space complexity solution
    vector<int> dailyTemperatures2(vector<int>& temperatures) {
        int n = temperatures.size();
        int hottest = 0;
        vector<int> res(temperatures.size(), 0);
        for (int i = n - 1; i >= 0; i--) {
            int curTemp = temperatures[i];
            if (curTemp >= hottest) {
                hottest = curTemp;
                continue;
            }
            int days = 1;
            while (temperatures[i + days] <= curTemp)
                days += res[i + days];
            res[i] = days;
        }
        return res;
    }
};